### ISSUE TITLE
Location Attention Memory Growth

### ISSUE THREAD

**Issue by PetrochukM** [pytorch/pytorch#7142]
## Issue description

A simple sample of Location Attention uses gigabytes of memory after 1000 iterations to store backward information. Using a similar attention model for audio generation, where 10 seconds of audio is approximately 800 iterations.

My investigation has led me to find that adding detach stems the problem; therefore, the memory consumed is due to gradient data. It is not clear to me how an alignment of size [64, 128] grows to 11 gigabytes of memory. An LSTM, a much more complicated unit, does not use this much memory after so many iterations.

Posted a sister thread here: https://discuss.pytorch.org/t/location-model-memory-growth/17304 

## Code example

```
import os

from torch import nn

import torch


class LocationAttention(nn.Module):
    """ Simple mock of a location sensative alignment similar to this paper:
    https://arxiv.org/pdf/1506.07503.pdf
    """

    def __init__(self, hidden_size):
        super().__init__()
        self.project_alignment = nn.Linear(1, hidden_size)
        self.score = nn.Linear(hidden_size, 1)
        self.softmax = nn.Softmax(dim=1)

    def forward(self, last_alignment):
        last_alignment = last_alignment.unsqueeze(2)
        last_alignment = self.project_alignment(last_alignment)
        score = self.score(last_alignment)
        score = score.squeeze(2)
        last_alignment = self.softmax(score)
        # last_alignment = last_alignment.detach()
        return last_alignment


# Parameters
batch_size = 64
num_tokens = 150
num_iter = 3000
hidden_size = 128

# Instantiate
attention = LocationAttention(hidden_size=hidden_size)
attention.cuda()

# Run
last_alignment = torch.FloatTensor(batch_size, num_tokens).zero_().cuda()
for i in range(num_iter):  # WARNING: Uses 11 gig of memory in 1600 iterations
    last_alignment = attention(last_alignment)
    if i % 100 == 0:
        print('Frame: %d' % (i,))
        os.system('nvidia-smi')
```

## System Info

PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudnn-7a90c013.so.7.0.5
/usr/local/lib/python3.5/dist-packages/torch/lib/libcudnn-3f9a723f.so.6.0.21

**Comment by apaszke** [comment]
It doesn't seem unexpected. You have two linear layers and 1600 steps. That's a lot of data to remember. Is there anything that makes you think that's a bug? Maybe try lowering the batch size?

**Comment by PetrochukM** [comment]
Thanks for getting in touch with me! Sorry if this issue is naive!

I'm faithfully replicating a paper from Google dealing with audio. It's common for a 10-second clip of audio to be represented by a sequence of 800 vectors. They are able to run their model without running out of memory on 10-second clips with a batch size of 64. 

I was expecting I would be able to do the same in PyTorch!

The demonstration above is that with a super simple Attention model PyTorch is running out of memory in 1600 steps. The real attention used is a bit more complicated in the paper. I run out of memory in 500 steps instead with it.

tl;dr I am hoping to replicate a paper from Google faithfully, that I expect was built in tensorflow but it seems that PyTorch is less memory efficient based on the evidence above. Noting that I have not worked much with tensorflow.

**Comment by apaszke** [comment]
The fact that your code doesn't match the (probably heavily tuned) TF code from Google doesn't necessarily mean that PyTorch is less memory efficient. There are many tricks like gradient checkpointing, or simply using multiple CUDA devices (with either data or model paralellism), that will let you fairly easily scale the example to longer sequences and larger batches. I'd be surprised if Google used a single GPU for that experiment, so that's likely where the difference lies. I can't see any immediate problems in your issue, so I'll close it, but if you'll manage to create simple PyTorch and TF snippets, where PyTorch is much less efficient than TF, then we'll be happy to investigate that. Currently we don't have anything to compare to. Hope you understand.

**Comment by PetrochukM** [comment]
@apaszke The paper clearly states that they use one GPU for this experiment. 

Where can I learn more about gradient checkpointing? What are other tricks? It's imperative for me to get this working.

I understand, thanks for looking into this briefly.

**Comment by soumith** [comment]
@PetrochukM where is the authors' code located, where is your code located?
gradient checkpointing docs are here: https://pytorch.org/docs/stable/checkpoint.html

**Comment by PetrochukM** [comment]
@soumith It'd be easier if they made their code public but they only have a paper out: https://arxiv.org/pdf/1712.05884.pdf 

Shared my code with you @soumith

**Comment by apaszke** [comment]
Another way to increase memory capacity 2x is to do the math in half precision.
